game-core/src/main/java/games/strategy/engine/data/GameData.java

```diff
@@ -215,8 +213,19 @@ public UnitTypeList getUnitTypeList() {
     return unitTypeList;
   }
 
-  public DelegateList getDelegateList() {
-    return delegateList;
+  public Collection<IDelegate> getDelegates() {
+    return delegates.values();
+  }
+
+  public void addDelegate(final IDelegate delegate) {
+    if (delegates == null) {
+      delegates = new HashMap<>();
+    }
+    delegates.put(delegate.getName(), delegate);
+  }
+
+  public IDelegate getDelegate(final String name) {
```

Comment:
This API could be a bit brittle:
- free-form String value, this creates magic constant, room for error, and is a more difficult API to use as one has to find the constant values
- breaks encapsulation, by knowing the 'key' for the delegates, client code knows about implementation details.

Was there a strong reason to not use an enum instead of `String`? Perhaps even better, any objection if we just enumerate the different methods and avoid casting? EG: `public PurchaseDelegate getPurchaseDelegate()`

Reply:
I mean the API was almost exactly the same before I did the smaller refactorings, so this is probably out-of-scope.
Regardless I'm afraid that there are probably too many Delegates to create methods for each (also the strings passed to this method are rarely hardcoded as far as I could tell, so this would end in giant switch cases).
Not sure about the enum approach. Might be an Idea worth thinking about, but I have no idea on how to implement this: Too much relies on this Delegate name. Using enums would only move the problem up one level and throw an exception when trying to convert an invalid string to an enum.

Reply:
Enum is a more appropriate type as it is more constrained/specific. `String` is a data type for free-form text. An analogy would be to use `int` when only prime values are allowed. This is a Stringly-typed anti-pattern.

My comment is suggesting that the refactor done here was not quite the right one. At minimum marking this method as 'deprecated' I think would be a good move so we know not to use the method.

> Regardless I'm afraid that there are probably too many Delegates to create methods for each

Perhaps we don't need one for each just yet. With `getDelegates` deprecated, we could remove this method and replace with the couple specific `get***Delegate` methods that are so far needed.

Reply:
Agree that we are moving a poor API, what's the best way to signal to future readers that this is a poor API? I suspect at minimum a TODO comment to suggest that we replace the `String` parameter or create specific methods is a good way to track that. Personally fixing this directly is the best way to go and not port this method to the new location and replace it directly.

Reply:
I just had a look:
Only a single non-test-code line calls `getDelegate` with a hardcoded string value: `ExportMenu:196`.
Other uses are:
1. Validate if the IDelegate name string is correct
2. Execute a generic game step
3. Test Code

I dug a little bit deeper to find the reason IDelegates are not directly being passed around:
It looks like IDelegates are never being sent over the network, only their name is being sent, so they are being re-created on the other side of the network.
See https://github.com/triplea-game/triplea/blob/4abe6b6b946f6523304754286fe895c16d58b95d/game-core/src/main/java/games/strategy/engine/framework/ServerGame.java#L543-L550